### ISSUE TITLE
Nested `<template>`s are invalid AMP

### ISSUE THREAD

**Issue by danielrozenberg** [ampproject/amphtml#16056]
In the ancient PR #12853 we finally allowed nested `<template>`s (issue #5811,) however somehow we never got around to removing it as a constrain in the validator. Basically, our docs are lying about nested templates :(

Weird, I thought I did that already!

I'm gonna create a PR for this soon :)

**Comment by Gregable** [comment]
How important is this. Templates nested in other templates are a tricky corner case for HTML parsing. Technically, I believe the contents of a `<template>` are not tags when parsing, but data, so there is no such thing as a nested template, there  is just the string `<template>` inside a template tag.

**Comment by danielrozenberg** [comment]
This was already done and fixed in the above mentioned PR and issue (and other PRs for following up on various bugs this caused.) This current issue is only about the validator rule

**Comment by choumx** [comment]
Browsers parse `<template>` contents into [DocumentFragments](https://developer.mozilla.org/en-US/docs/Web/API/DocumentFragment). https://html.spec.whatwg.org/#the-template-element

You may be thinking of mustache.js which handles its templates as strings.

**Comment by Gregable** [comment]
This was done in runtime. I'm asking how important is it to do this in validation? I'd prefer not to allow this in validation.

**Comment by danielrozenberg** [comment]
Some browsers don't support DocumentFragments, in which case the `<template>` contents is either a string, or regular DOM. We handle all these cases when we parse the template

**Comment by Gregable** [comment]
The validator parser is not a browser and must make sure that the document is safe prior to engagement of javascript runtime.

**Comment by choumx** [comment]
Not allowing this in validator isn't terrible since there's a workaround (extract template and reference by `template="<id>"` attribute).

> Technically, I believe the contents of a `<template>` are not tags when parsing, but data, so there is no such thing as a nested template, there is just the string `<template>` inside a template tag.

I'm not sure I understand. We must be parsing and validating contents of template tags for amp-mustache security (`b/25669598`). Do you mean the validator parser always treats nested templates as strings?

**Comment by choumx** [comment]
Blocked on validator/cache support for nested templates. Not critical at the moment.

**Comment by ampprojectbot** [comment]
This issue hasn't been  updated in awhile. @Gregable Do you have any updates?

**Comment by Gregable** [comment]
> Do you mean the validator parser always treats nested templates as strings?

Largely. IIRC, there are some security exceptions for preventing issues when the browser treats <template> as a <div> and there are some "linter-like" rules that aren't guaranteed. However, the validator guarantees nothing about js rendered templates.

I think I'm going to close this as the situation with nested templates is still a real challenge for non-browser parsers.
